nscplugin/src/main/scala-3/scala/scalanative/nscplugin/PrepNativeInteropLate.scala

```diff
@@ -48,6 +48,34 @@ class PostInlineNativeInterop extends PluginPhase {
         case ty => ty
   }
 
+  override def transformApply(tree: Apply)(using Context): Tree = {
+    val defnNir = this.defnNir
+    def dealiasTypeMapper = DealiasTypeMapper()
+
+    tree match
+      // fromScalaFunction[T1, R](fn) -> _fromScalaFunction(fn)
+      case Apply(TypeApply(fun, tArgs), args) if defnNir.CFuncPtr_fromScalaFunction.contains(fun.symbol) =>
+        val idx = defnNir.CFuncPtr_fromScalaFunction.indexOf(fun.symbol)
+        val transformed = defnNir._CFuncPtr_fromScalaFunction(idx)
+        val cls = defnNir.CFuncPtrNClass(idx)
+        val tys = tArgs.map(t => dealiasTypeMapper(t.typeOpt))
+        cpy
+          .Apply(tree)(
+            ref(transformed),
+            args
+          )
+          .withAttachment(NirDefinitions.NonErasedTypes, tys)
```

Comment:
We can add attachment to outer `Apply` without transformation of inner `TypeApply`. This combined with the fact that on Scala 2 it's not needed as well, would allow to eliminate additional `_fromScalaFunction` method. 
In Scala2 this transformation was unused, but in Scala3 we need to get original type and save it in attachment, because the target of a lambda might get replaced with adapted (boxed and erased) variant of a function. https://github.com/lampepfl/dotty/blob/abf9a25d5f27a41c11c11c11cf671ed3cb33d10d/compiler/src/dotty/tools/dotc/transform/Erasure.scala#L409-L439


```suggestion
      case appy @ Apply(TypeApply(fun, tArgs), _) if defnNir.CFuncPtr_fromScalaFunction.contains(fun.symbol) =>
        val tys = tArgs.map(t => dealiasTypeMapper(t.typeOpt))
        app
          .withAttachment(NirDefinitions.NonErasedTypes, tys)
```

Reply:
Thanks! `_fromScalaFunction` was a thing I don't like in this change and I'm glad we can get rid of that!
done https://github.com/scala-native/scala-native/pull/3270/commits/cc62ca2486fb2c7cc855440bdd709ebf45a99f32

>  but in Scala3 we need to get original type and save it in attachment, because the target of a lambda might get replaced with adapted (boxed and erased) variant of a function.

Oh, that's why we didn't need this in Scala2, while Scala3 does. That makes sense, thank you for the explanation!